### ISSUE TITLE
appendToBody property breaks date selection in SingleDatePicker when in a Material-UI Dialog

### ISSUE THREAD

**Issue by gilbsgilbs** [airbnb/react-dates#1550]
**react-dates version**

- react-dates@19.0.3

**Describe the bug**

When `appendToBody` is set in a `SingleDatePicker` and the date picker is in a Material-UI dialog, the picker displays correctly on click but `onDateChange` callback is never called and the picker gets closed immediately. The date doesn't update.

As I said in #1535 , #1540 has fixed the most the obvious cases (thanks again @danedavid for looking into this before I had a chance), yet some edge cases like this one still remain unfortunately.

**Source code (including props configuration)**

I was able to reproduce the issue by wrapping the picker in a Material-UI Dialog component:

```jsx
class DP extends Component {
  constructor(props) {
    super(props);
    this.state = {
      focused: props.autoFocus,
      date: props.initialDate,
    };

    this.onDateChange = this.onDateChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
  }

  onDateChange(date) {
    this.setState({ date });
  }

  onFocusChange({ focused }) {
    this.setState({ focused });
  }

  render() {
   const { focused, date } = this.state; 
    return (
        <SingleDatePicker
          id="mydatepicker"
          date={date}
          focused={focused}
          onDateChange={this.onDateChange}
          onFocusChange={this.onFocusChange}
          appendToBody
        />
    );
  }
}

class App extends Component {
  render() {
    return (
      <Dialog open>
        <DialogContent>
          <DP />
        </DialogContent>
      </Dialog>
    );
  }
}
```

See repro on CodeSandbox [here](https://codesandbox.io/s/nl9q7qzyj).

Would probably need some further investigations to understand why Material-UI's Dialogs produce such behaviors on React-Dates.

**Screenshots/Gifs**

None.

**Desktop (please complete the following information):**
* OS: Archlinux
* Browser: Chromium, Firefox
* Version: 72, 65

**Is the issue reproducible in Storybook?**

Not sure.

**Comment by gilbsgilbs** [comment]
Ok, it appears that this diff does fix the issue:

```diff
diff --git a/src/components/SingleDatePicker.jsx b/src/components/SingleDatePicker.jsx
index 0d631fa..9b9000d 100644
--- a/src/components/SingleDatePicker.jsx
+++ b/src/components/SingleDatePicker.jsx
@@ -248,7 +248,10 @@ class SingleDatePicker extends React.PureComponent {
 
   onFocusOut(e) {
     const { onFocusChange } = this.props;
-    if (this.dayPickerContainer.contains(e.relatedTarget || e.target)) return;
+    if (
+        (e.relatedTarget && this.dayPickerContainer.contains(e.relatedTarget))
+        || this.dayPickerContainer.contains(e.target)
+    ) return;
     onFocusChange({ focused: false });
   }
```

but I have to admit I have no clue whether this is correct or not (disclaimer: it's probably not).

Maybe @monokrome you could help me figure this out? It appears that in the case of the `Dialog`, `relatedTarget` is set to the dialog node and `target` is set to the calendar day node on the `FocusEvent`. I don't see a case where `relatedTarget` (the element that **gained** focus afaik?) could actually be contained by the picker container itself. So if there's none, then this is probably just enough:

```diff
diff --git a/src/components/SingleDatePicker.jsx b/src/components/SingleDatePicker.jsx
index 0d631fa..9b9000d 100644
--- a/src/components/SingleDatePicker.jsx
+++ b/src/components/SingleDatePicker.jsx
@@ -248,7 +248,10 @@ class SingleDatePicker extends React.PureComponent {
 
   onFocusOut(e) {
     const { onFocusChange } = this.props;
-    if (this.dayPickerContainer.contains(e.relatedTarget || e.target)) return;
+    if (this.dayPickerContainer.contains(e.target)) return;
     onFocusChange({ focused: false });
   }
```

In short: is `e.relatedTarget` really useful to look here?

Thanks.

**Comment by uvtzxpm** [comment]
I'm still having this exact problem with the current v20.3.0 version.

Did anything ever happen here? Maybe there's some other issue that I can't find?

I don't really want to go forking or monkey patching, so this is blocking me from upgrading past v18.4.0.

😭

**Comment by vlad-khitev-axon** [comment]
Still not working with v.21.5.0. Does anyone find a solution except rolling back to v18.4.0?

**Comment by gilbsgilbs** [comment]
I opened a #1890 since we weren't able to draw attention and get some help here.

**Comment by alfonmga** [comment]
I'm not even to make it works using `withPortal={true}` prop. 

I tried it also by overriding `DateRangePicker_picker__portal` styles and adding `z-index: 9999;` but it just doesn't work. It loses the focus and closes itself:

![2020-05-07 at 15 21](https://user-images.githubusercontent.com/9363272/81299465-8f1ffb80-9076-11ea-9b57-76d148fac109.gif)

I have been stuck with this issue for a few hours… do you guys know how to solve this? :/
